File: in_Python/0033 Search in Rotated Sorted Array.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        res = -1
        i, j = 0, len(nums)-1
        if len(nums) == 1:
            if nums[0] == target:
                return 0
            else:
                return -1
        while i < j:
            m = i + (j-i) / 2
            if nums[m] == target:
                return m
            if nums[i] == target:
                return i
            if nums[j] == target:
                return j
            if nums[j] > target:
                if nums[m] > nums[j]:
                    i = m + 1
                else:
                    if nums[m] < target:
                        i = m + 1
                    else:
                        j = m - 1
                continue
            elif nums[i] < target:
                if nums[m] < nums[i]:
                    j = m - 1
                else:
                    if nums[m] < target:
                        i = m + 1
                    else:
                        j = m - 1
                continue
            else:
                return -1
        return res

File: in_Java/0033 Search in Rotated Sorted Array.java
class Solution {
    public int search(int[] nums, int target) {
        if(nums == null || nums.length == 0) return -1;
        if(nums.length == 1){
            if(nums[0] == target) return 0;
            else return -1;
        }
        int i = 0; int j= nums.length-1;
        while(i <= j){
            int m = i + (j-i)/2;
            if(nums[m] == target) return m;
            if(nums[i] == target) return i;
            if(nums[j] == target) return j;
            if(nums[i] < nums[m]){
                if(target < nums[m] && target > nums[i]) j = m-1;
                else i = m + 1;
            }else{
                if(target > nums[m] && target < nums[j]) i = m + 1;
                else j = m - 1;
            }
    
        }
        return -1;
    }
}